074-additional-volumes-support.md

```diff
@@ -0,0 +1,95 @@
+# Support of additional volumes in pod
+
+We propose to enhance the Strimzi Kafka Operator by adding support for specifying additional volumes in the CRD. This will involve modifying the Kafka CRDs to include an `additionalVolumes` field, which will allow users to define extra volumes and volume mounts.
+
+The specific usecase of ours, is that we want to set custom log4j logging for audit logs, to ensure that audit logs are *not* present in stdout, but can be picked up by a sidecar which uses OpenTelemtry or logstash to forward to the audit logs to a secure location.
+
+## Current situation
+
+The operator handles many configurations and operational aspects but lacks support for specifying additional volumes in the Custom Resource Definition (CRD). This limitation can restrict users from accessing log files or integrating with external storage solutions directly.
+
+## Motivation
+
+Adding support for additional volumes in the CRD will allow users to:
+
+ - Access log files for enhanced monitoring and debugging.
+ - Use additional configurations or secret management tools that require mounting extra volumes.
+
+Improve flexibility in managing Kafka deployments by allowing users to attach custom storage solutions based on their specific requirements.
+
+## Proposal
+
+### Considerations
+
+- CRD Changes: Update the Kafka CRDs to include an `additionalVolumes` field.
+- Validation: Ensure that the additional volumes are validated correctly and do not conflict with existing volume definitions such as `/var/lib/kafka/` and `/tmp`.
+- Documentation: Update the Strimzi documentation to include examples and guidelines on how to use the new `additionalVolumes` field.
+- Security: Ensure that the additional volumes do not introduce security vulnerabilities by validating the configurations.

```

Comment:
i added some points around security in a comment below. I'm not sure if this can be validated or just have to be accepted as a risk. But it would be good if you can elaborate on what this means.

Reply:
Would this allow a route to running code that's not part of the container image, and therefore isn't immutable?

I guess I'm thinking of what's proposed here in conjunction with changing the `$CLASSPATH` to include a mounted directory and loading code/plugins from there? TBH I don't recall if we try to prevent users setting the `$CLASSPATH`, but even if we do, this would still be possible for Kafka Connect, because it uses a configurable class loading mechanism. 

I'm sure there are users who would like to be able to do this (and avoid the steps needed to include plugin jars in their own container images), but I'm not so sure that the project wants to support users who did this.

Reply:
@tombentley I agree that users shouldn't mount arbitrary code such as plugins using these custom volumes. So we should probably add a note in the documentation about this not being recommended

Reply:
> Would this allow a route to running code that's not part of the container image, and therefore isn't immutable?
> 
> I guess I'm thinking of what's proposed here in conjunction with changing the `$CLASSPATH` to include a mounted directory and loading code/plugins from there? TBH I don't recall if we try to prevent users setting the `$CLASSPATH`, but even if we do, this would still be possible for Kafka Connect, because it uses a configurable class loading mechanism.
> 
> I'm sure there are users who would like to be able to do this (and avoid the steps needed to include plugin jars in their own container images), but I'm not so sure that the project wants to support users who did this.

I share the concern about this being used to add runtime dependencies. But I'm not entirely convinced we should try to prevent it beyond discouraging it in the docs. It seems some users want to do it and the idea of playing some game where we try to plug the way to change the classpath with them seems like a waste of effort.

So I guess I would stick to docs and possibly enforce some mount path to avoid them remounting some existing binary directories directly etc.